oap-server/analyzer/log-analyzer/src/main/java/org/apache/skywalking/oap/log/analyzer/provider/log/listener/LogFilterListener.java

```diff
@@ -19,61 +19,76 @@
 import org.apache.skywalking.oap.log.analyzer.provider.LALConfigs;
 import org.apache.skywalking.oap.log.analyzer.provider.LogAnalyzerModuleConfig;
+import org.apache.skywalking.oap.server.core.UnexpectedException;
+import org.apache.skywalking.oap.server.core.analysis.Layer;
 import org.apache.skywalking.oap.server.library.module.ModuleManager;
 
 @Slf4j
 @RequiredArgsConstructor
 public class LogFilterListener implements LogAnalysisListener {
-    private final List<DSL> dsls;
+    private final DSL dsl;
 
     @Override
     public void build() {
-        dsls.forEach(dsl -> {
-            try {
-                dsl.evaluate();
-            } catch (final Exception e) {
-                log.warn("Failed to evaluate dsl: {}", dsl, e);
+        try {
+            if (dsl == null) {
+                log.warn("No rules match, the process will skip.");
+                return;
             }
-        });
+            dsl.evaluate();
+        } catch (final Exception e) {
+            log.warn("Failed to evaluate dsl: {}", dsl, e);
+        }
     }
 
     @Override
     public LogAnalysisListener parse(final LogData.Builder logData,
                                      final Message extraLog) {
-        dsls.forEach(dsl -> dsl.bind(new Binding().log(logData.build())
-                                                  .extraLog(extraLog)));
+        if (dsl == null) {
+            return null;
+        }
+        dsl.bind(new Binding().log(logData.build()).extraLog(extraLog));
         return this;
     }
 
     public static class Factory implements LogAnalysisListenerFactory {
-        private final List<DSL> dsls;
+        private final Map<Layer, DSL> dsls;
 
         public Factory(final ModuleManager moduleManager, final LogAnalyzerModuleConfig config) throws Exception {
-            dsls = new ArrayList<>();
+            dsls = new HashedMap<>();
 
             final List<LALConfig> configList = LALConfigs.load(config.getLalPath(), config.lalFiles())
                                                          .stream()
                                                          .flatMap(it -> it.getRules().stream())
                                                          .collect(Collectors.toList());
             for (final LALConfig c : configList) {
-                dsls.add(DSL.of(moduleManager, config, c.getDsl()));
+                try {
+                    Layer layer = Layer.nameOf(c.getLayer());
+                    dsls.put(layer, DSL.of(moduleManager, config, c.getDsl()));
+                } catch (UnexpectedException e) {
+                    log.warn("layer not found, will ignore this rule:" + c.getName());
+                }
             }
         }
 
         @Override
-        public LogAnalysisListener create() {
-            return new LogFilterListener(dsls);
+        public LogAnalysisListener create(Layer layer) {
+            if (layer == null) {
+                return null;
```

Comment:
Check the places where this method `create` is invoked, if you return null here, `NullPointerException` might be thrown, you have to check the nullability in the callers.

Reply:
I have checked it. No other invoker use the return value.

Reply:
> I have checked it. No other invoker use the return value.

Hi, `listeners` is an `ArrayList`, it doesn't allow null, so if `create` returns `null`, `NullPointerException` will be thrown in the following line

https://github.com/apache/skywalking/blob/81878331e59c435e011eb8c73fe5617201708968/oap-server/analyzer/log-analyzer/src/main/java/org/apache/skywalking/oap/log/analyzer/provider/log/LogAnalyzer.java#L79-L80

Reply:
@kezhenxu94 I have tested a null value situation. It will not throw a Exception.
An ArrayList can store more than one null.

Reply:
> @kezhenxu94 I have tested a null value situation. It will not throw a Exception.
> An ArrayList can store more than one null.

Hey, hosting a `null` value in the list is not a good idea. What's more, did you test the case where the logs will be persisted? The following lines when iterating the array list will throw exception. `listeners.forEach(LogAnalysisListener::build);` Element in `listeners` might be null and calling `null.build()` should throw.

https://github.com/apache/skywalking/blob/166a62612d20e403241aec24b95ceef1c64d898e/oap-server/analyzer/log-analyzer/src/main/java/org/apache/skywalking/oap/log/analyzer/provider/log/LogAnalyzer.java#L73-L81

Reply:
I understand , it's my fault. I will change it soon.